Add ability to subscribe to storage events

esimkowitz opened the pull request:
This PR updates the storage feature to a Signals-centric implementation that mirrors the features of the [Web Storage API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Storage_API). In this model, there are two main types of storage backing: LocalStorage (which is shared between all sessions) and SessionStorage (which is unique to each session). It also defines two types of StorageEntry, the application-side representation of the value in storage. 

The main StorageEntry struct contains mechanisms for reading to and updating from the underlying StorageBacking. Every StorageBacking can produce a StorageEntry struct. When initialized, a value will be fetched from storage if available and if the StorageEntry value ever changes, that value will be written back to storage. The SyncedStorageEntry adds to the StorageEntry struct a subscription mechanism. Only StorageBackings that also implement the StorageSubscriber trait can produce a SyncedStorageEntry struct. All SyncedStorageEntry structs for the same key will map to the same underlying storage value. If any of these instances updates the underlying value, all instances will be notified and will receive the new value. 

The actual mechanism for subscribing to these changes is platform-specific. For web targets, a listener is attached to the [Web API - Window - Storage Event](https://developer.mozilla.org/en-US/docs/Web/API/Window/storage_event). Whenever this changes, the new value is read per-session and all subscribing SyncedStorageEntry instances in that session are notified via a Tokio Watch channel. For desktop targets, all sessions are on the same thread so we can just use a common Tokio Watch channel for all windows, negating the need to watch the underlying storage.

The behavior of the SessionStorage StorageBacking is also different for different platforms. On the web, this uses the Web SessionStorage API, which ties to the current tab context. If a tab is duplicated, the SessionStorage gets cloned. On desktop, the SessionStorage is just an in-memory hashmap that is not forked if a new window is created.

This also adds two examples for the storage feature that show off the session storage and

esimkowitz committed:
save work

esimkowitz committed:
swap back to async_broadcast

esimkowitz committed:
remove comments

esimkowitz committed:
save work

esimkowitz committed:
save work

esimkowitz committed:
it works!

esimkowitz committed:
remove unnecessary dependency

esimkowitz commented:
While I've used use_persistent as my example here for subscription mainly because it was already implemented, I think it's actually a bad candidate for subscription. I think use_persistent should be adjusted so that it uses a session-specific store that expires and is cleaned up when the current session ends. I don't think there's a good use case for these values to persist between sessions, since this could easily lead to leakage of storage when the source code changes. Adding a function to purge persistent storage could help, but I feel like it's a burden to put this on the developer to remember to do.

I think instead we should align to the web API and create two StorageBacking impls per target - a LocalStorage that persists between sessions and supports synchronization, and a SessionStorage that is purged at the end of a session and does not support synchronization. I think that use_persistent falls more into the SessionStorage.

On desktop, it may just be easiest to represent SessionStorage as a static section of memory to ensure that it can be tied clearly to the application lifetime and eliminate any chance of storage leakage

esimkowitz committed:
remove unused imports

esimkowitz committed:
add localstorage

esimkowitz committed:
add unsubscribe

esimkowitz committed:
Merge branch 'storage-v4' into storage-v5

esimkowitz committed:
save work

esimkowitz committed:
change name

esimkowitz committed:
Break subscriber out

esimkowitz committed:
break out subscribe

esimkowitz committed:
update example

esimkowitz commented:
I've renamed ClientStorage in the web target to LocalStorage and added a second StorageBacking that points to SessionStorage. I've also added a separate trait for StorageSubscriber, so that storage mediums that don't (or shouldn't) support subscription (like SessionStorage) can just implement the read/write operations of StorageBacking. This also helps with correctness and setting developer expectation, since Rust will now fail to compile if you try to create a synced storage entry on a StorageBacking that doesn't implement StorageSubscriber. Still deciding if it is a good idea to also block creation of storage entries without a subscription if the StorageBacking _does_ support subscription (without this restriction, multiple instances could overwrite the same backing storage, though maybe that's fine).

I've updated the example too so that it has two different counts: one that persists for only the current session and one that persists for all sessions with synchronization.

I'm not super happy with the way that dereferencing now happens for the use_synced_storage_entry. I tried moving back to using UseRef but I ran into some borrowing issues. Do you have any pointers for this? 

I think my next order of business will be to swap the use of async_broadcast channels for [tokio::sync::watch](https://docs.rs/tokio/latest/tokio/sync/watch/index.html) and assigning one per key, rather than sharing. Watch channels allow for polling on the Sender closed condition, meaning I wouldn't need to explicitly unsubscribe keys from the watcher (this isn't relevant for web, since we only have one watcher on all events, but it will be for filesystem). I will hand all subscription and channel creation over to the watcher task and will use [oneshot channels](https://docs.rs/tokio/latest/tokio/sync/oneshot/index.html) to communicate the watch receiver back to the subscriber.

esimkowitz committed:
small refactoring

esimkowitz committed:
update sample

esimkowitz committed:
change UsePersistent to UseStorageEntry, use everywhere

esimkowitz committed:
more refactoring

ealmloff commented on src/storage/persistence.rs:
```diff
@@ -0,0 +1,73 @@
+        }
+        #[cfg(all(not(feature = "ssr"), not(feature = "hydrate")))]
+        {
+            use_signal(cx, || {
```
We can use signal here, but then we need to change the other flags to also use signals instead of use_ref. Currently the ssr feature is broken

esimkowitz replied:
yeah I focused on the one case I was currently testing for the prototype, will swap the rest out

esimkowitz replied:
How does SSR benefit from this feature? Curious if it's even worth implementing. Maybe we should just panic for SSR to avoid devs thinking it's doing something it's not. I could see this being useful for fullstack though (store the settings in the browser and communicate them back to the backend if needed)

ealmloff replied:
> How does SSR benefit from this feature? Curious if it's even worth implementing. Maybe we should just panic for SSR to avoid devs thinking it's doing something it's not. I could see this being useful for fullstack though (store the settings in the browser and communicate them back to the backend if needed)

SSR doesn't really use this feature, but we still need to provide a default value instead of panicing in SSR mode to make hydration work.

The approach I used initially did this:
On the server: Always use the value that should appear when storage is empty
On the client: First hydrate with the value that should appear when storage is empty on the first renderer, then update to the real value after hydration is finished

ealmloff commented on src/storage/storage_entry.rs:
```diff
@@ -0,0 +1,371 @@
+
+//  Start UseStorageEntry
+/// Storage that persists across application reloads
+pub struct UseStorageEntry<S: StorageBacking, T: Serialize + DeserializeOwned + Clone + 'static> {
```
Since this is backed by a signal we can implement Copy

esimkowitz replied:
fixed

ealmloff commented:
> I'm not super happy with the way that dereferencing now happens for the use_synced_storage_entry. I tried moving back to using UseRef but I ran into some borrowing issues. Do you have any pointers for this?

I wouldn't make it deref to the signal because the two structs have methods with the same name but different behavior:

```rust
*storage.write() = 0
// Doesn't do the same thing
*(&*storage).write() = 0
```

You could create an `inner` or `raw` method that exposes the inner signal instead

ealmloff commented on src/storage/storage_entry.rs:
```diff
@@ -0,0 +1,371 @@
+    }
+}
+
+pub struct StorageRef<'a, S: StorageBacking, T: Serialize + DeserializeOwned + Clone + 'static> {
```
Can this just be Ref instead of it's own type? That would allow users to use methods like [map](https://doc.rust-lang.org/std/cell/struct.Ref.html#method.map)

esimkowitz replied:
yeah i was thinking the same, will refactor

esimkowitz replied:
Trying to map the `Write<Signal<StorageEntry<S,T>>>` to `Write<T>`, but the dioxus-signals `Write::map` function is acting funky. I am trying the following map:

```rust
Write::map(self.inner.write(), |entry: &mut StorageEntry<S, T>| &mut entry.data)
```

However, I'm seeing an error when trying to return this, where it's saying that the above is actually returning Write<'_, _, StorageEntry<S,T>>. I looked at the source code for this and I see that the Write struct is actually `Write<'a, T: 'static, I: 'static = T>`. I am not sure this map can ever work because of the restriction that I = T, meaning the SignalSubscriberDrop and the underlying RefMut must always be the same type...

esimkowitz replied:
I've removed the StorageRef, etc. by swapping .read() for a Ref and removing write altogether. Instead of exposing a RefMut, I just expose .set() and .modify(). This also let me remove the StorageEntryMut struct too, since I can just use with_mut() to set and save the value

ealmloff replied:
> Trying to map the `Write<Signal<StorageEntry<S,T>>>` to `Write<T>`, but the dioxus-signals `Write::map` function is acting funky. I am trying the following map:
> 
> ```rust
> Write::map(self.inner.write(), |entry: &mut StorageEntry<S, T>| &mut entry.data)
> ```
> 
> However, I'm seeing an error when trying to return this, where it's saying that the above is actually returning Write<'_, _, StorageEntry<S,T>>. I looked at the source code for this and I see that the Write struct is actually `Write<'a, T: 'static, I: 'static = T>`. I am not sure this map can ever work because of the restriction that I = T, meaning the SignalSubscriberDrop and the underlying RefMut must always be the same type...

`I = T` just means that if I isn't specified then it can be assumed to be equal to `T`. If you make the function return `Write<'a, ORIGINAL_TYPE, NEW_TYPE>` then it should work

(related docs here https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#default-generic-type-parameters-and-operator-overloading)

esimkowitz replied:
i'll give this a try, thanks!

esimkowitz replied:
I was able to get this to work, by returning type `Write<'_, NEW_TYPE, ORIGINAL_TYPE>`, however I ran into an issue when dereferencing. I was able to resolve this by modifying the deref impl for Write to the following (basically just adding `I` constraint explicitly):

```rust
impl<'a, T: 'static, I: 'static> Deref for Write<'a, T, I> {
    type Target = T;

    fn deref(&self) -> &Self::Target {
        &self.write
    }
}

impl<T, I> DerefMut for Write<'_, T, I> {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.write
    }
}
```

ealmloff replied:
That is definitely an oversight. I generally rely on cargo to warn me if I am missing generics, but with default types both with and without the generic is valid. I wish there was a way to `#[warn(default_generic_impl)]`...

If you file a PR with your Dioxus fork I would be happy to merge it

esimkowitz replied:
I've opened a PR for this: https://github.com/DioxusLabs/dioxus/pull/1549

esimkowitz committed:
Remove custom UseStorageEntryRef

esimkowitz committed:
remove StorageEntryMut in favor of with_mut

esimkowitz committed:
add clone and copy derives

esimkowitz committed:
save work

esimkowitz committed:
potential signal-centric alternative

esimkowitz committed:
save

esimkowitz committed:
it works!

esimkowitz committed:
fix persistence

esimkowitz commented:
I tried simplifying the setup even more, not sure if I took it too far lol. Here's an implementation where the data inside StorageEntry is now stored in a signal, rather than storing the StorageEntry struct itself in a signal. This lets me just return back a vanilla Signal on deref. I use the subscriber model to then listen for changes and save them back to the storage. I think I could even decompose StorageEntry further with this, not sure if I should...

https://github.com/esimkowitz/dioxus-std/pull/1

ealmloff commented:
> I tried simplifying the setup even more, not sure if I took it too far lol. Here's an implementation where the data inside StorageEntry is now stored in a signal, rather than storing the StorageEntry struct itself in a signal. This lets me just return back a vanilla Signal on deref. I use the subscriber model to then listen for changes and save them back to the storage. I think I could even decompose StorageEntry further with this, not sure if I should...
> 
> [esimkowitz#1](https://github.com/esimkowitz/dioxus-std/pull/1)

I like that approach better because it exposes all the same methods as a signal!

esimkowitz committed:
use_effect instead of use_subscribe

esimkowitz committed:
fix use_effect

esimkowitz committed:
Merge pull request #1 from esimkowitz/storage-v4.1

Alternate storage sync implementation with deeper use of signals

esimkowitz committed:
working on ssr

esimkowitz committed:
refactoring for devex

esimkowitz committed:
add comments

esimkowitz committed:
add more comments

esimkowitz committed:
format

esimkowitz committed:
remove unnecessary constraints

esimkowitz committed:
change name

esimkowitz committed:
fix ssr

esimkowitz committed:
expose getters for channel and key in storageentry

esimkowitz committed:
format

ealmloff commented:
Is this PR ready?

esimkowitz commented:
The web stuff is ready to go. I still need to implement the subscribe for the filesystem. I got swamped this week with work, probably will have time to work on this on Monday. I have a pretty solid idea of how to implement that, I just haven't sat down to actually bang it out yet.

ealmloff commented:
> The web stuff is ready to go. I still need to implement the subscribe for the filesystem. I got swamped this week with work, probably will have time to work on this on Monday. I have a pretty solid idea of how to implement that, I just haven't sat down to actually bang it out yet.

Sounds good! No rush at all, I just wanted to make sure you were not waiting on me

esimkowitz committed:
update to use watch channels

esimkowitz committed:
fix cargo

esimkowitz committed:
simplify synced save

esimkowitz committed:
add more comments, fix unsubscribe

esimkowitz committed:
split storage desktop and web examples

esimkowitz committed:
Merge pull request #2 from esimkowitz/storage-v4.2

update to use watch channels

esimkowitz commented:
I've updated my implementation to use Watch channels instead of UseChannel. This lets me implement a stronger subscription mechanism, where each key gets its own channel. Now, I can fetch from storage once on an event and only notify the contexts that subscribe to the specific keys triggering the event. I can also more easily clean up old subscriptions, as I can watch for when the Sender gets closed (which happens automatically once all the Receivers are dropped) and then clean up any backing resources (like file watchers).

I've added a new trait for common StorageEntry functions and created a new wrapper struct SyncedStorageEntry to separate out the synchronization logic from the main struct even more.

I've also had to rework some of the internal hooks given this change to use a trait to encapsulate common StorageEntry functions.

esimkowitz committed:
remove semicolon

esimkowitz committed:
typo

esimkowitz committed:
remove dashmap due to wasm compatibility issues

esimkowitz committed:
add comments

esimkowitz committed:
adding filesystem and memory stores

esimkowitz commented:
The filesystem impl is now complete. I added a separate example for desktop. I want to add to this so that it demonstrates the multi-window support, then I think this will be ready

esimkowitz committed:
more comments

esimkowitz committed:
Updating storage desktop example to add new window

esimkowitz commented:
I updated the desktop example to support multiple windows. The syncing is working the first time the page comes into scope, but then it isn't resubscribing if you route away from the entry and return to it. It got me thinking though that since the desktop windows all run in the same thread, I can just use ThreadLocal channels to notify subscribers of changes and skip the file watcher altogether. This would be a lot cheaper, but would break if we ever change the desktop behavior to use separate threads per window

esimkowitz committed:
Remove notify crate

esimkowitz committed:
fix typo

esimkowitz committed:
Merge pull request #3 from esimkowitz/storage-v4.3

esimkowitz commented:
This is now fixed. I've verified that everything works as expected without the notify watcher.

esimkowitz committed:
add more comments

esimkowitz committed:
make log info into trace

esimkowitz commented:
I found a bug in the in-memory store. It is getting shared between windows since they are all on the same thread. I am going to see if I can tie it to the Dioxus context

esimkowitz commented:
This is my ideal approach to solve the above:

Make a hashtable that we encapsulate in an Arc and tie to the context lifetime using `cx.provide_root_context`. This has the advantage of cleaning up after itself when a scope is dropped, but it requires a rewrite of the StorageBacking trait to expose the context to it, even though it will only be used in one implementation.

Ideally, provide_root_context should be exposed as a global function similar to https://github.com/DioxusLabs/dioxus/pull/1572. This is already possible in dioxus-core using the with_runtime function, the same function that underpins provide_root_context.

esimkowitz commented:
Added a new issue for the above: https://github.com/DioxusLabs/dioxus/issues/1583

esimkowitz committed:
fix memory map

esimkowitz committed:
Update in-memory to use Any instead of serialize

esimkowitz commented:
I've fixed the in-memory store issue, thanks @ealmloff for the tip! With that, I think this PR is officially ready

ealmloff commented:
Thank you! There are some trait bound issues with the web example, but it seems pretty easy to fix, so I can fix those after I merge this
